napari/plugins/exceptions.py

```diff
@@ -1,23 +1,251 @@
     ) -> None:
         super().__init__(message)
         self.plugin_name = plugin_name
         self.plugin_module = plugin_module
+        PLUGIN_ERRORS[plugin_name].append(self)
+
+    def format_with_contact_info(self) -> str:
+        """Make formatted string with context and contact info if possible."""
+        # circular imports
+        from napari import __version__
+
+        msg = f'\n\nPluginError: {self}'
+        msg += '\n(Use "Plugins > Plugin errors..." to review/report errors.)'
+        if self.__cause__:
+            cause = str(self.__cause__).replace("\n", "\n" + " " * 13)
+            msg += f'\n  Cause was: {cause}'
+        contact = fetch_module_metadata(self.plugin_module)
+        if contact:
+            extra = [f'{k: >11}: {v}' for k, v in contact.items()]
+            extra += [f'{"napari": >11}: v{__version__}']
+            msg += "\n".join(extra)
+        msg += '\n'
+        return msg
+
+    def info(
+        self,
+    ) -> Tuple[Type['PluginError'], 'PluginError', Optional[TracebackType]]:
+        """Return info as would be returned from sys.exc_info()."""
+        return (self.__class__, self, self.__traceback__)
 
 
 class PluginImportError(PluginError, ImportError):
     """Raised when a plugin fails to import."""
 
     def __init__(self, plugin_name: str, plugin_module: str) -> None:
-        msg = f'Failed to import plugin: "{plugin_name}"'
+        msg = f"Failed to import plugin: '{plugin_name}'"
         super().__init__(msg, plugin_name, plugin_module)
 
 
 class PluginRegistrationError(PluginError):
     """Raised when a plugin fails to register with pluggy."""
 
     def __init__(self, plugin_name: str, plugin_module: str) -> None:
-        msg = f'Failed to register plugin: "{plugin_name}"'
+        msg = f"Failed to register plugin: '{plugin_name}'"
         super().__init__(msg, plugin_name, plugin_module)
+
+
+def format_exceptions(plugin_name: str, as_html: bool = False) -> str:
+    """Return formatted tracebacks for all exceptions raised by plugin.
+
+    This method takes advantage of ``IPython.core.ultratb.VerboseTB`` to create
+    a nicely formatted traceback.  If ``as_html`` is True, is then converted
+    from ansi to html.  (This approach was used instead of cgitb because the
+    IPython ultratb tracebacks were preferred).
+
+    Parameters
+    ----------
+    plugin_name : str
+        The name of a plugin for which to retrieve tracebacks.
+    as_html : bool
+        Whether to return the exception string as formatted html,
+        defaults to False.
+
+    Returns
+    -------
+    str
+        A formatted string with traceback information for every exception
+        raised by ``plugin_name`` during this session.
+    """
+    from napari import __version__
+
+    if not PLUGIN_ERRORS.get(plugin_name):
+        return ''
+
+    color = 'Linux' if as_html else 'NoColor'
+    vbtb = IPython.core.ultratb.VerboseTB(color_scheme=color)
```

Comment:
We are discussing in other issues making IPython an optional dependency, so I wonder if we can vendor this particular class to avoid depending on all of IPython just to colour tracebacks?

Reply:
at first glance, it looks like `VerboseTB` uses quite a bit of other things in IPython.  so, can we leave that for a later PR when we actually remove the dependency?  worst case scenario, we can just have them be uncolored...

Reply:
We made the import of IPython lazy in #1055 which sped up launch time quite a bit. Could we at least also make this import lazy here too? We could then make optional coloring in another PR

Reply:
@tlambert03 can you also elaborate on why this is so much better than cgitb? Can one be swapped for the other reasonably easily? If so, then we could make IPython lazy and optional here and that would just be wonderful. =) But, you can choose to leave it as just lazy and we can leave the optional part for a later PR.